### PR TITLE
feat: impl Into for Span/Array pair

### LINKED ISSUE #5631: feat: provide a `Span<T>` to `Array<T>` method
# Feature Request

`Span<T>` is a wrapper around `@Array<T>`. 
It should be possible to get the value out of this wrapper, either as `@Array<T>` or `Array<T>`. Or both.

Right now the only way to build back an `Array` from a `Span` is to iterate over it through `pop_front`. It is not a good DevX.

**Describe Preferred Solution**

A counterpart to the `Array::span()` method. `Span::consume -> Array` and `Span::snapshot -> @Array` maybe. 
Or implementations of the `Into` trait.

**Describe Alternatives**

 Implementations of the `Into` trait.

**Related Code**

[](https://github.com/starkware-libs/cairo/blob/68b7072783c7963e4ca17c88d9928dd5dc12db96/corelib/src/array.cairo#L4)

**If the feature request is approved, would you be willing to submit a PR?**
_(Help can be provided if you need assistance submitting a PR)_

No, it's an important design decision, it should most likely be done by the project maintainers.

### PR DESCRIPTION
**Pull request by tdelabro**
fixes #5631

- - -
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)

### CODE REVIEW

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: all files reviewed, 2 unresolved discussions (waiting on @tdelabro)

___
*[`corelib/src/array.cairo` line 151 at r1](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NytpHfkD574OY2dI-3m:-NytpHfkD574OY2dI-3n:b2dqiex) ([raw file](https://github.com/starkware-libs/cairo/blob/033caf2329406f2e2fdc06c985e961f6d3599827/corelib/src/array.cairo#L151)):*
> ```Cairo
>         self.span()
>     }
> }
> ```

there's not equivalent for making a vec into a span using into in rust.
please find out why before we decide on adding such a thing.



_Code quote:_
```Cairo
impl ArrayIntoSpan<T, +Drop<T>> of Into<Array<T>, Span<T>> {
    fn into(self: Array<T>) -> Span<T> {
        self.span()
    }
}
```
___
*[`corelib/src/array.cairo` line 161 at r1](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NytpxsUCjDQye2qMadi:-NytpxsUCjDQye2qMadj:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/033caf2329406f2e2fdc06c985e961f6d3599827/corelib/src/array.cairo#L161)):*
> ```Cairo
>         arr
>     }
> }
> ```





_Suggestion:_
```Cairo
impl SpanIntoArray<T, +Drop<T>, +Clone<T>> of Into<Span<T>, Array<T>> {
    fn into(mut self: Span<T>) -> Array<T> {
        let mut arr = array![];
        arr.append_span(self);
        arr
    }
}
```

**Review by tdelabro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: all files reviewed, 2 unresolved discussions (waiting on @orizi)

___
*[`corelib/src/array.cairo` line 151 at r1](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NytpHfkD574OY2dI-3m:-NyuGP5X31BFKM8UHvRw:b-tc2os9) ([raw file](https://github.com/starkware-libs/cairo/blob/033caf2329406f2e2fdc06c985e961f6d3599827/corelib/src/array.cairo#L151)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

there's not equivalent for making a vec into a span using into in rust.
please find out why before we decide on adding such a thing.
</blockquote></details>

tldr: there is not an exact equivalence between rust `&[]` and Cairo snapshots.

In Rust the implementation would not compile:
```rust
// input is passed by value
fn into(input: Vec<T>) -> &[T] {
   &input

  // `input` is dropped here
  // making any reference to it invalid due to lifetime issue
}
```

So there is no way to implement `From<Vec<T>> for &[]` in Rust.
The way to go from one to the other is through reference. Because `Vec<T>` implements `AsRef<[T]>`, `&Vec<T>` can be coerced into `&[T]`.

In Cairo, things are different because a snapshot can outlive the value it was created from.
Meaning you can do the following:
```cairo
let snap = my_vec.snap();
drop(my_vec);

// snap is still valid here because, due to the immutable memory, his lifetime is `'static`
snap.pop_next();
```

So an `Into` implementation is totally valid. It will create a snapshot of the array and drop the initial array in the same call. Nothing prohibited there. It's a valid type conversion, doing exactly what it advertises "consume an array and return a snapshot of it"

**Review by tdelabro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: 0 of 1 files reviewed, 2 unresolved discussions (waiting on @orizi)

___
*[`corelib/src/array.cairo` line 161 at r1](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NytpxsUCjDQye2qMadi:-NyuJJ7w61Tmt2XxpN-u:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/033caf2329406f2e2fdc06c985e961f6d3599827/corelib/src/array.cairo#L161)):*
> ```Cairo
>         arr
>     }
> }
> ```

Done.

**Review by orizi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @tdelabro)

___
*[`corelib/src/array.cairo` line 151 at r1](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NytpHfkD574OY2dI-3m:-NyuTkjJ8HNGqHFN-ARs:bd905wy) ([raw file](https://github.com/starkware-libs/cairo/blob/033caf2329406f2e2fdc06c985e961f6d3599827/corelib/src/array.cairo#L151)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

tldr: there is not an exact equivalence between rust `&[]` and Cairo snapshots.

In Rust the implementation would not compile:
```rust
// input is passed by value
fn into(input: Vec<T>) -> &[T] {
   &input

  // `input` is dropped here
  // making any reference to it invalid due to lifetime issue
}
```

So there is no way to implement `From<Vec<T>> for &[]` in Rust.
The way to go from one to the other is through reference. Because `Vec<T>` implements `AsRef<[T]>`, `&Vec<T>` can be coerced into `&[T]`.

In Cairo, things are different because a snapshot can outlive the value it was created from.
Meaning you can do the following:
```cairo
let snap = my_vec.snap();
drop(my_vec);

// snap is still valid here because, due to the immutable memory, his lifetime is `'static`
snap.pop_next();
```

So an `Into` implementation is totally valid. It will create a snapshot of the array and drop the initial array in the same call. Nothing prohibited there. It's a valid type conversion, doing exactly what it advertises "consume an array and return a snapshot of it"
</blockquote></details>

So conversion into a span should be done through something closer to the deref mechanism (currently being added, so probably not here)

**Review by tdelabro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @orizi)

___
*[`corelib/src/array.cairo` line 151 at r1](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NytpHfkD574OY2dI-3m:-NyukCY4AEbflAkglLdu:b-ko3fgi) ([raw file](https://github.com/starkware-libs/cairo/blob/033caf2329406f2e2fdc06c985e961f6d3599827/corelib/src/array.cairo#L151)):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

So conversion into a span should be done through something closer to the deref mechanism (currently being added, so probably not here)
</blockquote></details>

The fact that this is a totally valid way to create the object makes me think that it should be implemented.

Otherwise, it hurts the ability of the programmers to write useful generic traits. 

Any trait that has a `+Into<Span<T>>` bound should accept an `Array<T>` as a generic type, just for the sake of completeness. The opposite should be true too.

In Rust it doesn't make sense because references cannot outlive the value they are referencing. Cairo is entirely different in this regard and there is no reason to impair programmers for the sake of respecting constraints that only exist in another language.

Rust references and Cairo snapshots are just two different things, with different limitations, leading to different possibilities of usage.

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: all files reviewed, 1 unresolved discussion (waiting on @tdelabro)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5650#-Nz3ARh38K3GxVXYqLaU:-Nz3ARh4ETKdssD_oi6l:b-2xgzdi) (no related file):*
Add tests in array_test.cairo

___

**Review by tdelabro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: 1 of 2 files reviewed, 1 unresolved discussion (waiting on @orizi)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5650#-Nz3ARh38K3GxVXYqLaU:-Nz41OdbEb8bwJO_fuka:b-896fix) (no related file):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

Add tests in array_test.cairo
</blockquote></details>

Done.

___

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: 1 of 2 files reviewed, 3 unresolved discussions (waiting on @tdelabro)

___
*[`corelib/src/test/array_test.cairo` line 180 at r3](https://reviewable.io/reviews/starkware-libs/cairo/5650#-Nz43o9n1VWeuT7FaQG-:-Nz43o9n1VWeuT7FaQG0:bkhao89) ([raw file](https://github.com/starkware-libs/cairo/blob/5dcb9a2726a5c63dd4e6b4a5c3458e57932748ee/corelib/src/test/array_test.cairo#L180)):*
> ```Cairo
> }
> 
> mod into_trait_impls {
> ```

remove the outer module.
prefix all test names with `test_`.
___
*[`corelib/src/test/array_test.cairo` line 217 at r3](https://reviewable.io/reviews/starkware-libs/cairo/5650#-Nz4411-9SeR60EQSHBY:-Nz4411-9SeR60EQSHBZ:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/5dcb9a2726a5c63dd4e6b4a5c3458e57932748ee/corelib/src/test/array_test.cairo#L217)):*
> ```Cairo
> 
>         assert_eq!(array_snap, @array![1, 2, 3]); 
>     }
> ```





_Suggestion:_
```Cairo
    #[test]
    fn array_into_span() {
        assert_eq!(array![1, 2, 3].into(), array![1, 2, 3].span())
    }


    #[test]
    fn span_into_array() {
        assert_eq!(array![1, 2, 3].span().into(), array![1, 2, 3]);
    }

    #[test]
    fn array_snap_into_span() {
        assert_eq!((@array![1, 2, 3]).into(), array![1, 2, 3].span())
    }

    #[test]
    fn span_into_array_snap() {
        assert_eq!(array![1, 2, 3].span().into(), @array![1, 2, 3]);
    }
```

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @tdelabro)

___
*[`corelib/src/test/array_test.cairo` line 219 at r4](https://reviewable.io/reviews/starkware-libs/cairo/5650#-Nzb6e9J910McPLgQTYS:-Nzb6e9J910McPLgQTYT:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/d956f3ce5b04d03d7d7c870c24716316ee9dc027/corelib/src/test/array_test.cairo#L219)):*
> ```Cairo
> fn tets_span_into_array_snap() {
>     assert_eq!(@array![1, 2, 3], array![1, 2, 3].span().into()); 
> }
> ```





_Suggestion:_
```Cairo
fn test_array_into_span() {
    assert_eq!(array![1, 2, 3].span(), array![1, 2, 3].into())
}

#[test]
fn test_span_into_array() {
    assert_eq!(array![1, 2, 3], array![1, 2, 3].span().into());
}

#[test]
fn test_array_snap_into_span() {
    assert_eq!(array![1, 2, 3].span(), (@array![1, 2, 3]).into());
}

#[test]
fn test_span_into_array_snap() {
    assert_eq!(@array![1, 2, 3], array![1, 2, 3].span().into());
}
```

**Review by tdelabro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @orizi)

___
*[`corelib/src/test/array_test.cairo` line 219 at r4](https://reviewable.io/reviews/starkware-libs/cairo/5650#-Nzb6e9J910McPLgQTYS:-NzbXR358kU9-iJipfgN:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/d956f3ce5b04d03d7d7c870c24716316ee9dc027/corelib/src/test/array_test.cairo#L219)):*
> ```Cairo
> fn tets_span_into_array_snap() {
>     assert_eq!(@array![1, 2, 3], array![1, 2, 3].span().into()); 
> }
> ```

Done.

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: all files reviewed, 2 unresolved discussions (waiting on @tdelabro)

___
*[`corelib/src/array.cairo` line 161 at r5](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NzbaiTm4q2iSx16KlbY:-NzbaiTm4q2iSx16KlbZ:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/f63aee7e7dfb35bb0571d1ebf2558383069a9791/corelib/src/array.cairo#L161)):*
> ```Cairo
>         arr
>     }
> }
> ```





_Suggestion:_
```Cairo
impl SpanIntoArray<T, +Drop<T>, +Clone<T>> of Into<Span<T>, Array<T>> {
    fn into(self: Span<T>) -> Array<T> {
        let mut arr = array![];
        arr.append_span(self);
        arr
    }
}
```

**Review by orizi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: all files reviewed, 2 unresolved discussions (waiting on @tdelabro)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5650#-Nzbapqx0T1FviHfgUYy:-NzbapqyEIt9hPAAPhQb:b-uuj1x0) (no related file):*
@gilbens-starkware for 2nd eye.

___

**Review by tdelabro — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: all files reviewed, 2 unresolved discussions (waiting on @orizi)

___
*[`corelib/src/array.cairo` line 161 at r5](https://reviewable.io/reviews/starkware-libs/cairo/5650#-NzbaiTm4q2iSx16KlbY:-NzcOH7K47ZBq5uChkls:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/f63aee7e7dfb35bb0571d1ebf2558383069a9791/corelib/src/array.cairo#L161)):*
> ```Cairo
>         arr
>     }
> }
> ```

Done.

**Review by orizi — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650#-:-NzcOu2Z4kOTUW0qs-nO:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @tdelabro)

**Review by orizi — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: all files reviewed, 1 unresolved discussion (waiting on @tdelabro)

**Review by gilbens-starkware — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650#-:-NzcoDtRCxgXvp0cLG3v:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @tdelabro)

**Review by orizi — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5650)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @tdelabro)

**Comment by tdelabro** [comment]
@orizi fmt done, ready to merge
